sds: additional support for symlink-based key rotation.

htuch opened the pull request:
There are a few limitations in our existing support for symlink-based
key rotation:
- We don't atomically resolve symlinks, so a single snapshot might have
  inconsistent symlink resolutions for different watched files.
- Watches are on parent directories, e.g. for /foo/bar/baz on /foo/bar,
  which doesn't support common key rotation schemes were /foo/new/baz
  is rotated via a mv -Tf /foo/new /foo/bar.

The solution is to provide a structured WatchedDirectory for Secrets to
opt into when monitoring DataSources. SDS will used WatchedDirectory
to setup the inotify watch instead of the DataSource path. On update, it will
read key/cert twice, verifying file content hash consistency.

Risk level: Low (opt-in feature)
Testing: Unit and integration tests added.

Fixes #13663
Fixes #10979
Fixes #13370

Signed-off-by: Harvey Tuch <htuch@google.com>

htuch commented:
@mattklein123 @lizan @tsaarni can you take a look? This will address #13663, #10979 and #13370, I'd like to obtain consensus on the API before adding implementation and tests.

htuch commented on api/envoy/config/core/v3/base.proto:
```diff
@@ -323,6 +340,11 @@ message DataSource {
 
+    // Local filesystem data source watched for changes, e.g. by inotify on
+    // Linux. Not all *DataSource* instantiations support this, please consult
+    // documentation for the enclosing message.
```
I'm not super happy with this contextual awareness or the need to have whoever is processing the parent message worry about concerns like atomic symlink resolution. An alternative idea I had was to allow an idea of virtual filesystems to be established similar to runtime and indirect into them from a DataSource. But, this is quite a bit more complicated and I'm not sure who really cares about atomicity here other than key rotation in the current API.

lizan replied:
Yeah agreed, wondering what the implementation would look like as well. The other question is how this is going to be validated.

htuch replied:
The implementation is actually super simple, if you take a look at https://github.com/envoyproxy/envoy/blob/9ff2d695ff12512fb7c838fd8b7afc3e853d3e0c/source/common/secret/sds_api.cc#L58, you can see that the watch is already computed based on gathering up all the `DataSource`s and computing paths, here we just need to do figure out common roots and subdirectories. That's kind of why I'm inclined towards this approach over essentially generalizing runtime, which is a much bigger project.

What would sway me towards a more general virtual filesystem would be use cases beyond key rotation where atomicity/snapshots matters. A quick grep reveals that this doesn't seem to be the case in the current API, but I'd be keen to hear on what we might benefit from in the future.

In terms of validation, I'd plan on basic unit/integration tests to validate the functional aspects. For the atomicity of swap, I think we could add a virtual function to `SdsApi` that the file reload code calls into after each file read, in the unit test we could override with something that inject delay and allows us to perform additional rotations.

mattklein123 replied:
Should we still support `filename` above? Could everything be watched even if the user never actually moves it?

htuch replied:
This is an interesting idea, but I think the issue is that we need to add callbacks and behaviors at every site in Envoy that handles a `DataSource`. E.g. the LuaJIT filter, which might normally just be loading a file statically once. If you'd like to see this happen I can open a tracking issue for the various sites that would be missing this, but it's probably a fairly big task that requires per-subsystem integration and testing to get the refresh logic right. Even for TLS certs, the difficulty isn't watching files, it's dealing with dynamic updates of TLS on the listener objects (hence the SDS dynamic provider framework).

tsaarni replied:
Would there be any backward compatible way to replace `DataSource` with something like `WatchedDataSource` in those use cases that do support reloading?

htuch replied:
Yeah, that's a reasonable idea, we could deprecate the existing `DataSource` fields in the TLS certificate message, it's a bit of API churn but probably worth it from a clarity perspective.

tsaarni commented on api/envoy/config/core/v3/base.proto:
```diff
@@ -313,6 +313,23 @@ message HeaderMap {
+
+  // The filename (not including directory) of the watched resource.
+  string filename = 3 [(validate.rules).string = {min_len: 1}];
+}
```
I wonder, would the description be a bit clearer if reorganizing a bit e.g.

```protobuf
// A file that is watched for changes, e.g. by inotify on Linux. The full 
// file path is constructed by combining *watched_root*/*subdirectory*/*filename*.
// The watch only triggers on a move event occurring in *watched_root*.
// For example, if *watch_root* is */foo*, *subdirectory* is a symlink *bar* and
// *filename* is *baz*, ``ln -s <some_path> /foo/new && mv -Tf /foo/new /foo/bar`` 
// will trigger a reload of ``/foo/bar/baz``.
message WatchedFile {
  // Path to watch. 
  string watch_root = 1 [(validate.rules).string = {min_len: 1}];

  // An optional subdirectory containing the watched file.
  string subdirectory = 2;

  // The filename (not including directory) of the watched resource.
  string filename = 3 [(validate.rules).string = {min_len: 1}];
}
```

tsaarni commented on api/envoy/extensions/transport_sockets/tls/v3/common.proto:
```diff
@@ -191,6 +194,9 @@ message TlsSessionTicketKeys {
       [(validate.rules).repeated = {min_items: 1}, (udpa.annotations.sensitive) = true];
 }
 
+// All *DataSource*s in this message support *WatchedFile* when
```
Same as above: `crl` DataSource is not watched & reloaded today, but is the intention to add that now?

tsaarni commented on api/envoy/extensions/transport_sockets/tls/v3/common.proto:
```diff
@@ -128,6 +128,9 @@ message PrivateKeyProvider {
   }
 }
 
+// All *DataSource*s in this message support *WatchedFile* when
```
Just to double check: it is said all DataSources will be supporting the new WatchedFile but `password` and `ocsp_stable` are not watched & reloaded today? Is the intention to add those as well?

mattklein123 replied:
This is related to my question above. could we just make everything watched? Or is that too complicated? Otherwise will we fail in all places where it isn't documented and supported?

tsaarni reviewed (commented):
Regardless of working with only a subset of all DataSource use cases, I like this API.  

This will work nicely with Kubernetes symlink swap as well, by leaving `subdirectory` optional  :+1:

mattklein123 commented on api/envoy/config/core/v3/base.proto:
```diff
@@ -313,6 +313,23 @@ message HeaderMap {
 }
 
+// A file that is watched for changes, e.g. by inotify on Linux.
+message WatchedFile {
```
Should we use this in the filesystem runtime code and deprecate whatever is there now? This seems complicated enough that it would be good to have it in one place? This assumes that the functionality is basically the same.

htuch replied:
This is a watch on a file rather than a tree, the filesystem runtime has a bunch of stuff around overlays as well, not sure how much common ground is left..

mattklein123 commented on api/envoy/config/core/v3/base.proto:
```diff
@@ -313,6 +313,23 @@ message HeaderMap {
+  // trigger a reload of ``/foo/bar/baz``.
+  string watch_root = 1 [(validate.rules).string = {min_len: 1}];
+
+  // An optional subdirectory containing the watched file.
```
Can you clarify whether this can be multiple sub-directories with the OS specific path splitter between them? Or just a single one?

htuch replied:
I could go either way on this; it would be simplest to make it a single subdirectory, but if we want the versatility here of arbitrary structure under the symlink, which we might in other potential use cases, then we can make it multiple. The implementation difference is that we need to do something like `realpath(3)` instead of `readlink(2)`. On Windows I think this path canonicalization is quite complex (CC @wrowe).

mattklein123 reviewed (commented):
At a high level LGTM with a few comments.

htuch commented:
I've updated the API to avoid modifying `DataSource`, introducing a new concept of an opt-in `WatchedDirectory`. I've also added a draft implementation and a very basis integration test. I'd like to achieve agreement on API and design, at which point I can clean up the implementation and fill out the rest of the tests. @lizan @mattklein123 @tsaarni PTAL.

htuch commented on include/envoy/filesystem/filesystem.h:
```diff
@@ -112,12 +112,29 @@ class Instance {
+   *         the supplied path is returned.
+   * @throw EnvoyException if the path can't be resolved.
+   */
+  virtual std::string readSymlink(absl::string_view path) PURE;
```
@wrowe would it be reasonable to land this on Linux/Mac and tag as unsupported on Windows, leaving a tracking issue for @envoyproxy/windows-dev? I think for new platform level APIs it would be good to have the Windows experts own these.

htuch replied:
Also, see the other comments I tagged you on, we might want this to be full blown path canonicalization, similar to the APR example you shared.

htuch replied:
Not needed any more since I switched to hash-based atomic reads.

tsaarni commented on source/common/config/watched_directory.cc:
```diff
@@ -0,0 +1,43 @@
+  try {
+    const auto result = filesystem_.splitPathFromFilename(data_source.filename());
+    if (result.directory_ == literal_path_) {
+      data_source.set_filename(filesystem_.joinPath(resolved_path_, result.file_));
```
Since filename is set according to `resolved_path_`, does it mean that `data_source.filename()` will change each iteration according to realpath? Will `result.directory_ == literal_path_` still evaluate true the next round?

I was wondering: is the `subdirectory` parameter necessary at all in `WatchedDirectory`? Would it be reasonable to make watched directory completely independent concept from `DataSource.filename`?  That is, filename would be used for reading files, watched directory for the watch?

htuch replied:
Updated, I think both concerns are addressed.

markdroth commented on api/envoy/config/core/v3/base.proto:
```diff
@@ -313,6 +313,17 @@ message HeaderMap {
+  // A subdirectory from *root* containing watched files. This will be resolved
+  // via readlink(2) of *root*/*subdirectory* on an inotify event prior to
+  // reading files.
+  string subdirectory = 2;
```
As we discussed offline, it seems a little strange to me to have to specify the parent directory and the subdirectory separately.  It seems like it should be possible to just directly monitor the subdirectory (which ideally would work regardless of whether or not it's a symlink).

htuch replied:
Updated to an interface providing more control.

htuch commented:
This now uses hash-based comparison of files after reading twice and validated consistency. This is similar to what gRPC is doing.

tsaarni reviewed (commented):
Looks good to me after the unbound look in `SdsApi::onWatchUpdate()` is resolved.

htuch commented on source/common/secret/sds_api.h:
```diff
@@ -223,19 +260,45 @@ class CertificateValidationContextSdsApi : public SdsApi,
+        envoy::extensions::transport_sockets::tls::v3::CertificateValidationContext>(
+        *sds_certificate_validation_context_secrets_);
+    // We replace path based secrets with inlined secrets on update.
+    if (watched_path_ != nullptr) {
```
I've moved it to unconditionally read from the filesystem, I think this is safe, the only place that I think path is really used by the TLS code is around certificate validation chain PEM.

htuch commented on source/common/filesystem/posix/filesystem_impl.cc:
```diff
@@ -131,6 +131,29 @@ PathSplitResult InstanceImplPosix::splitPathFromFilename(absl::string_view path)
+std::string InstanceImplPosix::readSymlink(absl::string_view path) {
+  std::string null_terminated_path{path};
+  char resolved_path[PATH_MAX];
+  const ssize_t result = ::readlink(null_terminated_path.data(), resolved_path, PATH_MAX);
```
Yep, walked this part of the PR back.

markdroth commented on api/envoy/config/core/v3/base.proto:
```diff
@@ -313,6 +313,24 @@ message HeaderMap {
 }
 
+// A path that is watched for changes, e.g. by inotify on Linux.
+message WatchedPath {
```
This looks much nicer!

Would it make sense to split this up into two separate messages, `WatchedDirectory` and `WatchedFile`, so that the user doesn't need to specify the event type?  That might be a more ergonomic way of expressing the same thing.

htuch replied:
Switched back to WatchedDirectory, I think that's all we need per @tsaarni.

mattklein123 commented on api/envoy/config/core/v3/base.proto:
```diff
@@ -313,6 +313,24 @@ message HeaderMap {
+    MODIFIED = 1;
+  }
+
+  // Path to watch.
```
Can you beef this up to clarify that this should be a path to a file? Or is the intention to support directory watching also? Per @markdroth comment about maybe splitting the messages and increasing the docs would help?

tsaarni replied:
I wonder is the file modify variant necessary or could it be dropped in favor of directory move watch only? 

Capability of watching modify is of course nice for completeness, and thanks to the new hashing approach, it probably has a good chance on returning consistent files, but shouldn't symlink/directory swap still be recommended?  Also, the `WatcherImpl` for inotify will set watch for directory regardless of user asking for file modify events, so there is little difference between the two on Linux.

htuch replied:
Yep, back to WatchedDirectory.

mattklein123 commented on api/envoy/extensions/transport_sockets/tls/v3/common.proto:
```diff
@@ -128,16 +128,31 @@ message PrivateKeyProvider {
+  // applies to dynamic secrets, when the *TlsCertificate* is delivered via SDS.
   config.core.v3.DataSource private_key = 2 [(udpa.annotations.sensitive) = true];
 
+  // If specified, updates of file-based *certificate_chain* and *private_key*
+  // sources will be triggered by this watch. The certificate/key pair will be
+  // read together and validated for atomic read consistency (i.e. no intervening
+  // modification occurred between cert/key read).
+  config.core.v3.WatchedPath watched_path = 7;
```
Probably related to my comments above but I'm confused as to whether we are watching a directory or a file here. Maybe different messages and more comments would help?

mattklein123 reviewed (changes requested):
Thanks agreed this seems like a better direction. Flushing out a few API questions/comments. 

/wait

htuch commented:
PTAL, if this makes sense now in terms of API and hash-based comparison implementation then I can fill in the missing tests/docs and move this out of draft.

mattklein123 commented on api/envoy/extensions/transport_sockets/tls/v3/common.proto:
```diff
@@ -128,16 +128,31 @@ message PrivateKeyProvider {
   config.core.v3.DataSource private_key = 2 [(udpa.annotations.sensitive) = true];
 
+  // If specified, updates of file-based *certificate_chain* and *private_key*
+  // sources will be triggered by this watch. The certificate/key pair will be
+  // read together and validated for atomic read consistency (i.e. no intervening
+  // modification occurred between cert/key read).
```
How is this actually going to be done? Is this an implementation detail that you don't want to document? It might matter to the operator though. Timestamp? Something else?

htuch replied:
Yeah, the plan is to make this hash-based, since we are already reading file contents in for hashing for other purposes (deciding whether to trigger a TLS update callback).

mattklein123 reviewed (changes requested):
From an API perspective this LGTM with one question. LMK when ready for review and I can take a look.

/wait

htuch commented:
@mattklein123 @lizan this is now ready for review.

mattklein123 commented on api/envoy/extensions/transport_sockets/tls/v3/common.proto:
```diff
@@ -233,8 +249,17 @@ message CertificateValidationContext {
+  // delivered via SDS.
   config.core.v3.DataSource trusted_ca = 1;
 
+  // If specified, updates of a file-based *trusted_ca* source will be triggered
+  // by this watch.
+  config.core.v3.WatchedDirectory watched_directory = 11;
```
What is the utility of this if there is only a single file being watched? Can this be removed for now for simplicity? Or is it that this works for non-SDS? If that's the case can you beef up the docs for both fields to make this more clear? It's kind of confusing.

htuch replied:
This allows you to watch at arbitrary points in the file path. E.g. for `/foo/bar/baz/cert.pem`, you can watch on `/foo/bar`. I'll work on improving docs.

mattklein123 commented on api/envoy/extensions/transport_sockets/tls/v3/common.proto:
```diff
@@ -128,16 +128,32 @@ message PrivateKeyProvider {
+  // applies to dynamic secrets, when the *TlsCertificate* is delivered via SDS.
   config.core.v3.DataSource private_key = 2 [(udpa.annotations.sensitive) = true];
 
+  // If specified, updates of file-based *certificate_chain* and *private_key*
+  // sources will be triggered by this watch. The certificate/key pair will be
+  // read together and validated for atomic read consistency (i.e. no intervening
+  // modification occurred between cert/key read, verified by file hash
+  // comparisons).
```
So this works for non-SDS file based certificates? Related to my comment below but if that's the case, can you beef up the docs here and above so they refer to each other a bit and make clear the difference? It's kind of hard to understand why one would use the different options here.

mattklein123 reviewed (changes requested):
LGTM with some API comments, thank you!

/wait

htuch commented:
@mattklein123 do deal with failure scenarios I had to add some stats, which then forced me to fix SDS stats more generally, since they were broken before (they all ended up anonymously in the root namespace). PTAL and LMK if there's anything else needed around stats here.

mattklein123 commented on api/envoy/extensions/transport_sockets/tls/v3/common.proto:
```diff
@@ -233,8 +253,21 @@ message CertificateValidationContext {
+  // delivered via SDS.
   config.core.v3.DataSource trusted_ca = 1;
 
+  // If specified, updates of a file-based *trusted_ca* source will be triggered
+  // by this watch. This allows explicit control over the path watched, by
+  // default the parent directory of the filesystem path in *trusted_ca* is
+  // watched if this field is not specified. This only applies when a
+  // *CertificateValidationContext* is delivered by SDS with references to
+  // filesystem paths.
+  config.core.v3.WatchedDirectory watched_directory = 11;
```
I still don't understand how this does anything if there is only a single file above? What's the utility?

htuch replied:
Let's say you have a trusted CA files `/foo/bar/baz/ca.pem`. By default, envoy watches `/foo/bar/baz`. If you want to reload when `baz` is atomically moved, you need to set the watch on `/foo/bar`. This is what this field allows.

mattklein123 replied:
OK I think having a concrete example might help for this stuff (or ref link back to the docs where you have the example) as it was pretty hard for me to understand the difference.

mattklein123 commented on source/common/secret/sds_api.cc:
```diff
@@ -45,28 +51,35 @@ void SdsApi::resolveDataSource(const FileContentMap& files,
-    files_hash_ = new_hash;
+  // Filesystem reads and update callbacks can fail if the key material is missing or bad. We're not
+  // under an onConfigUpdate() context, so we need to catch these cases explicitly here.
+  try {
```
Is it possible to have the scope of the try/catch be a bit smaller? Especially not the callbacks? NBD if not but just a thought.

htuch replied:
It needs to cover the callbacks, since they may throw exceptions, and the rotation happens outside of an `onConfigUpdate` when it's safe to do so.

mattklein123 replied:
OK makes sense.

mattklein123 reviewed (commented):
Thanks LGTM with small questions.

/wait-any

mattklein123 reviewed (approved):
LGTM. Feel free to do any doc updates as a follow up.

htuch commented:
Sure, will merge as some folks are depending on this, but I'll take an AI to do the docs followups next week. Thanks.
